tests/h/views/accounts_test.py

```diff
@@ -688,97 +688,71 @@ def routes(self, pyramid_config):
-        controller = views.NotificationsController(pyramid_request)
-        controller.form = invalid_form()
+    def test_get_if_user_has_no_email_address(self, controller, authenticated_user):
+        authenticated_user.email = None
 
-        result = controller.post()
+        result = controller.get()
 
-        assert "form" not in result
-        assert result["user_has_email_address"] is None
+        assert result == {"user_has_email_address": None}
 
-    def test_post_with_valid_data_updates_subscriptions(
-        self, form_validating_to, pyramid_config, pyramid_request, subscriptions_model
-    ):
-        pyramid_request.POST = {}
-        pyramid_config.testing_securitypolicy("fiona")
+    def test_post(self, controller, form_validating_to, subscriptions_model):
         subs = [FakeSubscription("reply", True), FakeSubscription("foo", False)]
         subscriptions_model.get_subscriptions_for_uri.return_value = subs
-        controller = views.NotificationsController(pyramid_request)
         controller.form = form_validating_to({"notifications": {"foo"}})
 
-        controller.post()
+        result = controller.post()
 
         assert not subs[0].active
         assert subs[1].active is True
+        # This appears to be testing `h.form.handle_form_submission()`
+        assert isinstance(result, httpexceptions.HTTPFound)
 
-    def test_post_with_valid_data_redirects(
-        self, form_validating_to, pyramid_config, pyramid_request, subscriptions_model
-    ):
-        pyramid_request.POST = {}
-        pyramid_config.testing_securitypolicy("fiona")
-        subscriptions_model.get_subscriptions_for_uri.return_value = []
-        controller = views.NotificationsController(pyramid_request)
-        controller.form = form_validating_to({})
+    def test_post_with_invalid_data(self, controller, authenticated_user, invalid_form):
+        authenticated_user.email = None
+        controller.form = invalid_form()
 
         result = controller.post()
 
-        assert isinstance(result, httpexceptions.HTTPFound)
+        assert result == {"user_has_email_address": None}
 
     @pytest.fixture
+    def controller(self, pyramid_request, form_validating_to):
+        controller = views.NotificationsController(pyramid_request)
+        controller.form = form_validating_to({})
+
+        return controller
+
+    @pytest.fixture(autouse=True)
+    def authenticated_user(self, pyramid_config, pyramid_request, factories):
+        authenticated_user = factories.User(email="email@example.com")
+        pyramid_config.testing_securitypolicy(userid=authenticated_user.userid)
+        pyramid_request.user = authenticated_user
+
+        return authenticated_user
+
+    @pytest.fixture(autouse=True)
     def routes(self, pyramid_config):
         pyramid_config.add_route("account_notifications", "/p/notifications")
 
+    @pytest.fixture(autouse=True)
+    def subscriptions_model(self, patch):
+        return patch("h.models.Subscriptions")
```

Comment:
No other tests use this fixture. In a test file 900 lines long this is really unhelpful to split up. It means surfing hundreds of lines to try and see the relevant fixtures.

It also makes it a pain if you come to move this into separate files at some point. If the tests are more self contained, you can lift and shift them with much less drama.